src/supervisor.rs

```diff
@@ -0,0 +1,614 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::context::{self, Context};
+use crate::events::EventNotifier;
+use crate::worker::{self, Worker};
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Order {
+    LeftToRight,
+    RightToLeft,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub enum Strategy {
+    OneForOne,
+}
+
+#[derive(Debug)]
+pub struct StartError {
+    pub failing_child_runtime_name: String,
+    pub failing_child_error: Arc<worker::StartError>,
+    pub failing_sibling_termination_error: Option<TerminationError>,
```

Comment:
Since `TerminatiorError` is a `HashMap` you could drop the `Option` here.  All the Rust standard library collections allocate lazily on first insert (unless you reserve space `with_capacity`), so an empty collection is a very cheap type and it fundamentally represents the same idea as an `Option<Collection>`.

Reply:
For some reason I thought it was better to detect there was a termination error via a `match` of `None`, rather than a `is_empty` call. I'll do this change 👍

Reply:
Yeah, I have that same impulse too. The thing that has started to bother me about it is that then there are two ways for it to be "empty" a `None` and an empty collection.

Reply:
Thinking about this a bit more, the details of the `HashMap` are embedded in the `TerminationError` type, so clients cannot perform an `is_empty` unless we make the `HashMap` attribute public or we offer a intermediary method that calls its `is_empty`. 

If we have control on how values of this type get created, we can enforce that whenever there is a `TerminationError` value, it will always have a full `HashMap`. Is there a way to make sure that API clients cannot create a value of this type, but can use one?

Reply:
That's a good point. As long as the inner field is not marked `pub` and you don't create a `pub` constructor method that takes a `HashMap` you'll be able to enforce how `TerminationError` values are created to always have a non-empty hash map.  In order to allow API users to access to the inner data you'll have to provide a public API that lets the consumer get an iterator or queries the hash-map directly, not sure what makes the most sense. Maybe both.

Reply:
To be clear, definitely fine to keep this field as an `Option<TerminationError>`.

Reply:
I'll do that for now, when the time comes, we will offer a way to see the termination errors on workers on a way that make sense on the API (no pesky is_empty checks)